### PR TITLE
[BEAM-11688] Support partial proto encoding

### PR DESCRIPTION
**Pull request by udim**
Refactored out common test values for coders that handle multiple types.

Joint work with @katsiapis.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by udim** [comment]
R: @ibzib 
CC: @robertwb

File: sdks/python/apache_beam/coders/coders_test_common.py [hunk 1]
```diff
@@ -63,6 +65,34 @@ class CodersTest(unittest.TestCase):
   # These class methods ensure that we test each defined coder in both
   # nested and unnested context.
 
+  # Common test values representing Python's built-in types.
```
Comment (ibzib) [inline sdks/python/apache_beam/coders/coders_test_common.py:68]: 👍

File: sdks/python/apache_beam/coders/coders_test_common.py [hunk 2]
```diff
@@ -63,6 +65,34 @@ class CodersTest(unittest.TestCase):
...
+  values_deterministic: List[Any] = [
```
Comment (ibzib) [inline sdks/python/apache_beam/coders/coders_test_common.py:69]: Nit: `values_deterministic` is kind of vague, `values` even more so. Can we rename to `test_values*`?

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 3]
```diff
@@ -336,18 +336,18 @@ def __init__(self, proto_message_type):
     self.proto_message_type = proto_message_type
 
   def encode(self, value):
-    return value.SerializeToString()
+    return value.SerializePartialToString()
 
   def decode(self, encoded):
     proto_message = self.proto_message_type()
-    proto_message.ParseFromString(encoded)
+    proto_message.ParseFromString(encoded)  # This is in effect "ParsePartial".
```
Comment (ibzib) [inline sdks/python/apache_beam/coders/coder_impl.py:343]: `# This is in effect "ParsePartial".` -> Why is this the case?

File: sdks/python/apache_beam/coders/coders_test_common.py [hunk 4]
```diff
@@ -129,12 +159,20 @@ def test_custom_coder(self):
...
+    coder = coders.PickleCoder()
+    self.check_coder(coder, *self.values)
 
   def test_deterministic_coder(self):
     coder = coders.FastPrimitivesCoder()
     deterministic_coder = coders.DeterministicFastPrimitivesCoder(coder, 'step')
-    self.check_coder(deterministic_coder, 'a', 1, 1.5, (1, 2, 3))
+    self.check_coder(deterministic_coder, *self.values_deterministic)
+    for v in self.values_deterministic:
+      self.check_coder(coders.TupleCoder((deterministic_coder, )), (v, ))
```
Comment (ibzib) [inline sdks/python/apache_beam/coders/coders_test_common.py:170]: What makes TupleCoder a special case here? What coverage does this check provide that `self.check_coder(deterministic_coder, (1, 2, 3))` didn't?

File: sdks/python/apache_beam/coders/coder_impl.py (hunk 3 above)
Comment (udim) [reply sdks/python/apache_beam/coders/coder_impl.py:343]: This comment is saying that this method can parse both initialized and partial messages. (there is no ParsePartialFromString)

File: sdks/python/apache_beam/coders/coders_test_common.py (hunk 4 above)
Comment (udim) [reply sdks/python/apache_beam/coders/coders_test_common.py:170]: The coverage is mainly for TupleCoder I believe? I don't know the reasoning but I wanted to stay consistent with the other tests in this file.

File: sdks/python/apache_beam/coders/coders_test_common.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/coders/coders_test_common.py]: done

**Comment by udim** [comment]
R: @robertwb

**Review by ibzib — APPROVED** [review]
LGTM

**Comment by aaltay** [comment]
Previous python precommit fail looks like a flake: https://ci-beam.apache.org/job/beam_PreCommit_Python_Commit/17052/

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by udim** [comment]
Run Python PreCommit

**Pull request merged**